api/server.go

```diff
@@ -1,28 +1,44 @@
 package api
 
 import (
+	"context"
 	"fmt"
 	"net/http"
 	"time"
 
 	"github.com/sirupsen/logrus"
 
-	"go.k6.io/k6/api/common"
 	v1 "go.k6.io/k6/api/v1"
-	"go.k6.io/k6/core"
+	"go.k6.io/k6/execution"
+	"go.k6.io/k6/lib"
+	"go.k6.io/k6/metrics"
+	"go.k6.io/k6/metrics/engine"
 )
 
-func newHandler(logger logrus.FieldLogger) http.Handler {
+func newHandler(cs *v1.ControlSurface) http.Handler {
 	mux := http.NewServeMux()
-	mux.Handle("/v1/", v1.NewHandler())
-	mux.Handle("/ping", handlePing(logger))
-	mux.Handle("/", handlePing(logger))
+	mux.Handle("/v1/", v1.NewHandler(cs))
+	mux.Handle("/ping", handlePing(cs.RunState.Logger))
+	mux.Handle("/", handlePing(cs.RunState.Logger))
 	return mux
 }
 
 // GetServer returns a http.Server instance that can serve k6's REST API.
-func GetServer(addr string, engine *core.Engine, logger logrus.FieldLogger) *http.Server {
-	mux := withEngine(engine, newLogger(logger, newHandler(logger)))
+func GetServer(
+	runCtx context.Context, addr string, runState *lib.TestRunState,
+	samples chan metrics.SampleContainer, me *engine.MetricsEngine, es *execution.Scheduler,
+) *http.Server {
+	// TODO: reduce the control surface as much as possible? For example, if
+	// we refactor the Runner API, we won't need to send the Samples channel.
```

Comment:
Out of curiosity, do we have an issue with it? I didn't find much searching by `Runner`.

Reply:
Hmm I don't think we have a dedicated issue for that :thinking: We have a few code TODOs like this one, and some tangentially related issues like https://github.com/grafana/k6/issues/1048 and https://github.com/grafana/k6/issues/2869, but nothing dedicated :thinking: 

The problem is that we (or at least I) don't have a clear vision for how it should refactored yet... For example, I know that we need to change the name, since `Runner` is super confusing when it is essentially a "factory" for VUs... :sweat_smile: But since I don't have a clear idea how we can refactor it, I also don't know into what to rename it :disappointed: 

So yeah, this is why we have the `TODO` here and in the `Runner`: https://github.com/grafana/k6/blob/1d99b0be9194bf9aaf86c070dd0d34604f266abe/lib/runner.go#L41-L85

I also have some very vague ideas on how to refactor setup/teardown/handleSummary handling so they don't require their own dedicated methods, and also how to refactor the archive handling to be somewhat separate, but everything is super vague at the moment and probably full of problems :disappointed: Someone needs to start doing it, so we can hit the real issues.

Reply:
Decided to create a new issue after all: https://github.com/grafana/k6/issues/2873

Even though we don't have the whole picture, it doesn't hurt to have a place to collect problems that need to be fixed